client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -330,6 +342,24 @@ private void fetchUpdatesIfNeeded() {
             compactIfNeeded();
         }
     }
+
+    void updateConfigIfNeeded() {
+        if (!updateConfigTimer.hasRemaining()) {
+            fetchUpdatesIfNeeded();
+            ReaderGroupState state = sync.getState();
+            if (state.isUpdatingConfig()) {
+                sync.updateState((s, updates) -> {
+                    ReaderGroupConfig controllerConfig = getThrowingException(controller.getReaderGroupConfig(scope, groupName));
+                    log.debug("Updating the readergroup {} with the new config {} obtained from the controller", groupName, controllerConfig);
+                    if (s.getConfig().getGeneration() < controllerConfig.getGeneration()) {
+                        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, controllerConfig);
```

Comment:
This is making multiple calls to the controller inside of the updateState function. This doesn't work well for optimistic concurrency. Perhaps most this outside of the update state, and just rely on the `if` to make the update conditional on the config not having changed.

Reply:
Moved the controller calls outside the update state.